src/frontend/src/handler/create_sink.rs

```diff
@@ -181,6 +182,19 @@ pub async fn gen_sink_plan(
         },
     };
 
+    // only allow Kafka connector work with `bytes` as key encode
+    if let Some(format_desc) = &format_desc
+        && let Some(encode) = &format_desc.key_encode
+        && connector != KAFKA_SINK
+        && matches!(encode, SinkEncode::Bytes)
+    {
+        return Err(ErrorCode::BindError(format!(
+            "key encode bytes only works with kafka connector, but found {}",
+            connector
+        ))
+        .into());
+    }
+
```

Comment:
It would be VERY HARD to locate and update the corresponding logic when the conditions change, if we keep *adding random `if`-checks here and there*. For this one, let's put it inside `fn validate_compatibility`.